File: admin/src/main/java/com/beat/admin/promotion/application/service/command/AdminPromotionCommandService.java
package com.beat.admin.promotion.application.service.command;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.beat.admin.common.application.converter.AdminCarouselNumberEnumConverter;

import com.beat.admin.promotion.application.dto.request.AdminCarouselNumber;
import com.beat.admin.promotion.application.dto.request.CarouselHandleRequest;
import com.beat.admin.promotion.application.dto.request.CarouselHandleRequest.PromotionGenerateRequest;
import com.beat.admin.promotion.application.dto.request.CarouselHandleRequest.PromotionModifyRequest;
import com.beat.admin.promotion.application.dto.request.PromotionHandleRequest;
import com.beat.admin.promotion.application.dto.response.CarouselHandleAllResponse;
import com.beat.admin.promotion.application.dto.result.AdminPromotionResults;
import com.beat.admin.promotion.application.dto.result.AdminPromotionResults.AdminPromotionResult;
import com.beat.admin.application.exception.AdminApplicationErrorCode;
import com.beat.domain.member.repository.MemberRepository;
import com.beat.domain.performance.repository.PerformanceRepository;
import com.beat.domain.promotion.domain.CarouselNumber;
import com.beat.domain.promotion.domain.Promotion;
import com.beat.domain.promotion.repository.PromotionRepository;
import com.beat.global.support.exception.BadRequestException;
import com.beat.global.support.exception.NotFoundException;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AdminPromotionCommandService {

	private static final Comparator<Promotion> BY_CAROUSEL_NUMBER = Comparator.comparing(
		Promotion::getCarouselNumber,
		Comparator.comparingInt(Enum::ordinal)
	);

	private final MemberRepository memberRepository;
	private final PromotionRepository promotionRepository;
	private final PerformanceRepository performanceRepository;

	@Transactional
	public CarouselHandleAllResponse processAllPromotionsSortedByCarouselNumber(Long memberId,
		CarouselHandleRequest request) {
		validateMemberExists(memberId);

		ClassifiedCarouselPromotions classifiedPromotions = classifyCarouselPromotions(request);

		List<Promotion> allExistingPromotions = promotionRepository.findAll();
		List<Long> deletePromotionIds = extractDeletePromotionIds(allExistingPromotions,
			classifiedPromotions.requestPromotionIds());
		List<Promotion> changedPromotions = processPromotions(classifiedPromotions.modifyRequests(),
			classifiedPromotions.generateRequests(), deletePromotionIds);

		return CarouselHandleAllResponse.from(
			toPromotionResults(changedPromotions)
		);
	}

	private AdminPromotionResults toPromotionResults(List<Promotion> domainPromotions) {
		List<AdminPromotionResult> promotionResults = domainPromotions.stream()
			.sorted(BY_CAROUSEL_NUMBER)
			.map(this::toPromotionResult)
			.toList();
		return AdminPromotionResults.from(promotionResults);
	}

	private AdminPromotionResult toPromotionResult(Promotion domainPromotion) {
		return AdminPromotionResult.of(
			domainPromotion.getId(),
			AdminCarouselNumberEnumConverter.toApiName(domainPromotion.getCarouselNumber()),
			domainPromotion.getPromotionPhoto(),
			domainPromotion.isExternal(),
			domainPromotion.getRedirectUrl(),
			domainPromotion.getPerformanceId()
		);
	}

	private void validateCarouselHandleRequest(CarouselHandleRequest request) {
		if (request == null || request.carousels() == null) {
			throw new BadRequestException(AdminApplicationErrorCode.INVALID_REQUEST_FORMAT);
		}
	}

	private ClassifiedCarouselPromotions classifyCarouselPromotions(CarouselHandleRequest request) {
		validateCarouselHandleRequest(request);

		List<PromotionModifyRequest> modifyRequests = new ArrayList<>();
		List<PromotionGenerateRequest> generateRequests = new ArrayList<>();
		Set<Long> requestPromotionIds = new HashSet<>();

		for (PromotionHandleRequest promotionRequest : request.carousels()) {
			switch (promotionRequest) {
				case PromotionModifyRequest modifyRequest -> {
					modifyRequests.add(modifyRequest);
					requestPromotionIds.add(modifyRequest.promotionId());
				}
				case PromotionGenerateRequest generateRequest -> generateRequests.add(generateRequest);
				case null, default -> throw new BadRequestException(AdminApplicationErrorCode.INVALID_REQUEST_FORMAT);
			}
		}

		return new ClassifiedCarouselPromotions(
			List.copyOf(modifyRequests),
			List.copyOf(generateRequests),
			new HashSet<>(requestPromotionIds)
		);
	}

	private List<Long> extractDeletePromotionIds(List<Promotion> allExistingPromotions, Set<Long> requestPromotionIds) {
		Set<Long> allExistingPromotionIds = allExistingPromotions.stream()
			.map(Promotion::getId)
			.collect(Collectors.toSet());

		return allExistingPromotionIds.stream()
			.filter(existingId -> !requestPromotionIds.contains(existingId))
			.toList();
	}

	private List<Promotion> processPromotions(List<PromotionModifyRequest> modifyRequests,
		List<PromotionGenerateRequest> generateRequests, List<Long> deletePromotionIds) {
		handlePromotionDeletion(deletePromotionIds);
		List<Promotion> modifiedDomainPromotions = handlePromotionModification(modifyRequests);
		List<Promotion> addedPromotions = handlePromotionGeneration(generateRequests);

		List<Promotion> appliedDomainPromotionChanges = new ArrayList<>(modifiedDomainPromotions);
		appliedDomainPromotionChanges.addAll(addedPromotions);

		return appliedDomainPromotionChanges;
	}

	private void handlePromotionDeletion(List<Long> deletePromotionIds) {
		if (!deletePromotionIds.isEmpty()) {
			promotionRepository.deleteByPromotionIds(deletePromotionIds);
		}
	}

	private List<Promotion> handlePromotionModification(List<PromotionModifyRequest> modifyRequests) {
		return modifyRequests.stream()
			.map(modifyRequest -> {
				Promotion promotion = findPromotionById(modifyRequest.promotionId());
				Long performanceId = validatePerformanceId(modifyRequest.performanceId());

				Promotion updatedPromotion = promotion.updatePromotionDetails(
					toCarouselNumber(modifyRequest.carouselNumber()),
					modifyRequest.newImageUrl(), modifyRequest.isExternal(), modifyRequest.redirectUrl(),
					performanceId);
				return promotionRepository.save(updatedPromotion);
			})
			.toList();
	}

	private List<Promotion> handlePromotionGeneration(List<PromotionGenerateRequest> generateRequests) {
		return generateRequests.stream()
			.map(generateRequest -> {
				Long performanceId = validatePerformanceId(generateRequest.performanceId());

				Promotion newPromotion = Promotion.create(generateRequest.newImageUrl(), performanceId,
					generateRequest.redirectUrl(), generateRequest.isExternal(),
					toCarouselNumber(generateRequest.carouselNumber()));
				return promotionRepository.save(newPromotion);
			})
			.toList();
	}

	private CarouselNumber toCarouselNumber(AdminCarouselNumber carouselNumber) {
		if (carouselNumber == null) {
			return null;
		}
		return AdminCarouselNumberEnumConverter.toDomain(carouselNumber);
	}

	private Promotion findPromotionById(Long promotionId) {
		return promotionRepository.findById(promotionId)
			.orElseThrow(() -> new NotFoundException(AdminApplicationErrorCode.PROMOTION_NOT_FOUND));
	}

	private Long validatePerformanceId(Long performanceId) {
		if (performanceId == null) {
			return null;
		}
		performanceRepository.findById(performanceId)
			.orElseThrow(() -> new NotFoundException(AdminApplicationErrorCode.PERFORMANCE_NOT_FOUND));
		return performanceId;
	}

	private void validateMemberExists(Long memberId) {
		memberRepository.findById(memberId)
			.orElseThrow(() -> new NotFoundException(AdminApplicationErrorCode.MEMBER_NOT_FOUND));
	}

	private record ClassifiedCarouselPromotions(
		List<PromotionModifyRequest> modifyRequests,
		List<PromotionGenerateRequest> generateRequests,
		Set<Long> requestPromotionIds
	) {
	}
}

File: application/admin/src/main/kotlin/com/beat/application/admin/promotion/command/AdminPromotionCommandService.kt
package com.beat.application.admin.promotion.command

import com.beat.application.admin.exception.AdminApplicationException
import com.beat.application.admin.exception.translateDomainFailure
import com.beat.application.admin.promotion.AdminPromotionResultAssembler
import com.beat.application.admin.promotion.AdminPromotionResults
import com.beat.application.admin.promotion.command.CarouselHandleCommand.PromotionGenerateCommand
import com.beat.application.admin.promotion.command.CarouselHandleCommand.PromotionModifyCommand
import com.beat.application.admin.promotion.exception.PromotionApplicationErrorCode
import com.beat.domain.member.repository.MemberRepository
import com.beat.domain.performance.repository.PerformanceRepository
import com.beat.domain.promotion.model.CarouselNumber
import com.beat.domain.promotion.model.Promotion
import com.beat.domain.promotion.repository.PromotionRepository
import com.beat.domain.promotion.service.PromotionCarouselDomainService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AdminPromotionCommandService
internal constructor(
    private val memberRepository: MemberRepository,
    private val promotionRepository: PromotionRepository,
    private val performanceRepository: PerformanceRepository,
    private val promotionImageCache: PromotionImageCache,
    private val promotionImageStorage: PromotionImageStorage,
    private val promotionCarouselDomainService: PromotionCarouselDomainService,
) {
    @Transactional
    fun processAllPromotionsSortedByCarouselNumber(
        memberId: Long,
        command: CarouselHandleCommand,
    ): AdminPromotionResults {
        return translateDomainFailure {
            validateMemberExists(memberId)

            val classifiedPromotions = classifyCarouselPromotions(command)
            validateCarouselAssignments(classifiedPromotions)
            validateCarouselImageObjects(classifiedPromotions)
            lockReferencedPerformances(classifiedPromotions)

            val allExistingPromotions = promotionRepository.lockAll()
            val deletePromotionIds =
                extractDeletePromotionIds(
                    allExistingPromotions,
                    classifiedPromotions.requestPromotionIds,
                )
            val changedPromotions =
                processPromotions(
                    classifiedPromotions.modifyRequests,
                    classifiedPromotions.generateRequests,
                    deletePromotionIds,
                )

            AdminPromotionResultAssembler.assemble(changedPromotions)
        }
    }

    private fun validateCarouselImageObjects(promotions: ClassifiedCarouselPromotions) {
        promotions.modifyRequests.forEach { validateCarouselImageObject(it.newImageUrl) }
        promotions.generateRequests.forEach { validateCarouselImageObject(it.newImageUrl) }
    }

    private fun validateCarouselImageObject(imageUrl: String) {
        val imageKey = requireNotNull(ImageKeyExtractor.extract(imageUrl))
        if (!promotionImageStorage.exists(imageKey)) {
            throw AdminApplicationException(PromotionApplicationErrorCode.INVALID_IMAGE_UPLOAD)
        }
    }

    private fun lockReferencedPerformances(promotions: ClassifiedCarouselPromotions) {
        val performanceIds =
            (promotions.modifyRequests.mapNotNull { it.performanceId } +
                    promotions.generateRequests.mapNotNull { it.performanceId })
                .distinct()
                .sorted()
        performanceIds.forEach { performanceId ->
            performanceRepository.lockById(performanceId)
                ?: throw AdminApplicationException(
                    PromotionApplicationErrorCode.PERFORMANCE_NOT_FOUND
                )
        }
    }

    private fun classifyCarouselPromotions(
        command: CarouselHandleCommand
    ): ClassifiedCarouselPromotions {
        val modifyRequests = command.carousels.filterIsInstance<PromotionModifyCommand>()
        val generateRequests = command.carousels.filterIsInstance<PromotionGenerateCommand>()

        // null이거나 sealed 하위 타입이 아닌 항목은 두 필터에서 모두 빠지므로 개수 합으로 감지
        if (modifyRequests.size + generateRequests.size != command.carousels.size) {
            throw AdminApplicationException(PromotionApplicationErrorCode.INVALID_REQUEST_FORMAT)
        }

        val requestPromotionIds = modifyRequests.map { it.promotionId }
        if (requestPromotionIds.distinct().size != requestPromotionIds.size) {
            throw AdminApplicationException(PromotionApplicationErrorCode.INVALID_REQUEST_FORMAT)
        }

        return ClassifiedCarouselPromotions(
            modifyRequests = modifyRequests,
            generateRequests = generateRequests,
            requestPromotionIds = requestPromotionIds.toSet(),
        )
    }

    private fun validateCarouselAssignments(promotions: ClassifiedCarouselPromotions) {
        val carouselNumbers =
            promotions.modifyRequests.map { toCarouselNumber(it.carouselNumber) } +
                promotions.generateRequests.map { toCarouselNumber(it.carouselNumber) }

        if (!promotionCarouselDomainService.hasValidCarouselAssignments(carouselNumbers)) {
            throw AdminApplicationException(PromotionApplicationErrorCode.INVALID_REQUEST_FORMAT)
        }
    }

    private fun extractDeletePromotionIds(
        allExistingPromotions: List<Promotion>,
        requestPromotionIds: Set<Long>,
    ): List<Long> {
        val allExistingPromotionIds = allExistingPromotions.map { requireNotNull(it.id) }.toSet()
        return allExistingPromotionIds.filter { it !in requestPromotionIds }
    }

    private fun processPromotions(
        modifyRequests: List<PromotionModifyCommand>,
        generateRequests: List<PromotionGenerateCommand>,
        deletePromotionIds: List<Long>,
    ): List<Promotion> {
        handlePromotionDeletion(deletePromotionIds)
        val modifiedDomainPromotions = handlePromotionModification(modifyRequests)
        val addedPromotions = handlePromotionGeneration(generateRequests)
        return modifiedDomainPromotions + addedPromotions
    }

    private fun handlePromotionDeletion(deletePromotionIds: List<Long>) {
        if (deletePromotionIds.isNotEmpty()) {
            promotionRepository.deleteByPromotionIds(deletePromotionIds)
        }
    }

    private fun handlePromotionModification(
        modifyRequests: List<PromotionModifyCommand>
    ): List<Promotion> = modifyRequests.map { modifyRequest ->
        val promotion = findPromotionById(modifyRequest.promotionId)
        val imageKey = requireNotNull(ImageKeyExtractor.extract(modifyRequest.newImageUrl))
        val updatedPromotion =
            promotion.updatePromotionDetails(
                carouselNumber = toCarouselNumber(modifyRequest.carouselNumber),
                newImageUrl = imageKey,
                isExternal = modifyRequest.isExternal,
                redirectUrl = modifyRequest.redirectUrl,
                performanceId = modifyRequest.performanceId,
            )
        promotionRepository.save(updatedPromotion).also { promotionImageCache.preWarm(imageKey) }
    }

    private fun handlePromotionGeneration(
        generateRequests: List<PromotionGenerateCommand>
    ): List<Promotion> = generateRequests.map { generateRequest ->
        val imageKey = requireNotNull(ImageKeyExtractor.extract(generateRequest.newImageUrl))
        val newPromotion =
            Promotion.create(
                promotionPhoto = imageKey,
                performanceId = generateRequest.performanceId,
                redirectUrl = generateRequest.redirectUrl,
                isExternal = generateRequest.isExternal,
                carouselNumber = toCarouselNumber(generateRequest.carouselNumber),
            )
        promotionRepository.save(newPromotion).also { promotionImageCache.preWarm(imageKey) }
    }

    private fun toCarouselNumber(carouselNumber: String): CarouselNumber =
        CarouselNumber.valueOf(carouselNumber)

    private fun findPromotionById(promotionId: Long): Promotion =
        promotionRepository.findById(promotionId)
            ?: throw AdminApplicationException(PromotionApplicationErrorCode.PROMOTION_NOT_FOUND)

    private fun validateMemberExists(memberId: Long) {
        memberRepository.findById(memberId)
            ?: throw AdminApplicationException(PromotionApplicationErrorCode.MEMBER_NOT_FOUND)
    }

    private data class ClassifiedCarouselPromotions(
        val modifyRequests: List<PromotionModifyCommand>,
        val generateRequests: List<PromotionGenerateCommand>,
        val requestPromotionIds: Set<Long>,
    )
}
